src/drivers/bmm150/bmm150.cpp

```diff
@@ -0,0 +1,1122 @@
+	}
+
+	/* allocate basic report buffers */
+	_reports = new ringbuffer::RingBuffer(2, sizeof(_report));
+
+	if (_reports == nullptr) {
+		goto out;
+	}
+
+	/* Bring the device to sleep mode */
+	modify_reg(BMM150_POWER_CTRL_REG, 1, 1);
+	up_udelay(10000);
+
+
+	/* check  id*/
+	if (read_reg(BMM150_CHIP_ID_REG) != BMM150_CHIP_ID) {
+		warnx("id of magnetometer is not: 0x%02x", BMM150_CHIP_ID);
+		return -EIO;
+	}
+
+	if (reset() != OK) {
+		goto out;
+	}
+
+	init_trim_registers();
+
+	_class_instance = register_class_devname(MAG_BASE_DEVICE_PATH);
+
+	collect();
+
+	/* advertise sensor topic, measure manually to initialize valid report */
+	struct mag_report mrb;
+	_reports->get(&mrb);
+
+	/* measurement will have generated a report, publish */
+	_topic = orb_advertise_multi(ORB_ID(sensor_mag), &mrb,
+				     &_orb_class_instance, (is_external()) ? ORB_PRIO_HIGH : ORB_PRIO_MAX);
+
+	if (_topic == nullptr) {
+		warnx("ADVERT FAIL");
+	}
+
+out:
+	return ret;
+}
+
+int
+BMM150::probe()
+{
+	/* During I2C Initialization, sensor is in suspend mode
+	 * and chip Id will return 0x00, hence returning OK. After
+	 * I2C initialization, sensor is brought to sleep mode
+	 * and Chip Id is verified. In sleep mode, we can read
+	 * chip id. */
+
+	/* @Note: Please read BMM150 Datasheet for more details */
+	return OK;
+}
+
+
+void
+BMM150::start()
+{
+	/* reset the report ring and state machine */
+	_reports->flush();
+
+	/* schedule a cycle to start things */
+	work_queue(HPWORK, &_work, (worker_t)&BMM150::cycle_trampoline, this, 1);
+
+
+}
+
+void
+BMM150::stop()
+{
+	work_cancel(HPWORK, &_work);
+}
+
+ssize_t
+BMM150::read(struct file *filp, char *buffer, size_t buflen)
+{
+	unsigned count = buflen / sizeof(_report);
+	struct mag_report *mag_buf = reinterpret_cast<struct mag_report *>(buffer);
+	int ret = 0;
+
+	/* buffer must be large enough */
+	if (count < 1) {
+		return -ENOSPC;
+	}
+
+	/* if automatic measurement is enabled */
+	if (_call_interval > 0) {
+		/*
+		 * While there is space in the caller's buffer, and reports, copy them.
+		 * Note that we may be pre-empted by the workq thread while we are doing this;
+		 * we are careful to avoid racing with them.
+		 */
+		while (count--) {
+			if (_reports->get(mag_buf)) {
+				ret += sizeof(struct mag_report);
+				mag_buf++;
+			}
+		}
+
+		/* if there was no data, warn the caller */
+		return ret ? ret : -EAGAIN;
+	}
+
+	/* manual measurement - run one conversion */
+	/* XXX really it'd be nice to lock against other readers here */
+	do {
+		_reports->flush();
+
+		/* run the collection phase */
+		if (OK != collect()) {
+			ret = -EIO;
+			break;
+		}
+
+
+		if (_reports->get(mag_buf)) {
+			ret = sizeof(struct mag_report);
+		}
+	} while (0);
+
+	/* return the number of bytes transferred */
+	return ret;
+
+}
+
+
+void
+BMM150::cycle_trampoline(void *arg)
+{
+	BMM150 *dev = reinterpret_cast<BMM150 *>(arg);
+
+	/* make measurement */
+	dev->cycle();
+
+}
+
+void
+BMM150::cycle()
+{
+	collect();
```

Comment:
When a driver runs off of the worker queue or as a task. There can be a race on stop and cycle. Please add a bool to the class and set it false in stop(), then wrap the

```
if (_running) {
  work_queue(HPWORK, &_work, (worker_t)&BMM150::cycle_trampoline, this, ....));
}
 ```  
so that that any pending cycles callbacks do not reschedule again.